vectorstores/pinecone/pinecone.go

```diff
@@ -0,0 +1,157 @@
+package pinecone
+
+import (
+	"context"
+	"errors"
+
+	"github.com/tmc/langchaingo/embeddings"
+	"github.com/tmc/langchaingo/schema"
+	"github.com/tmc/langchaingo/vectorstores"
+	"github.com/tmc/langchaingo/vectorstores/pinecone/internal/grpcapi"
+	"github.com/tmc/langchaingo/vectorstores/pinecone/internal/restapi"
+)
+
+var (
+	// ErrMissingTextKey is returned in SimilaritySearch if a vector
+	// from the query is missing the text key.
+	ErrMissingTextKey             = errors.New("missing text key in vector metadata")
+	ErrEmbedderWrongNumberVectors = errors.New(
+		"number of vectors from embedder does not match number of documents",
+	)
+)
+
+// Store is a wrapper around the pinecone rest API and grpc client.
+type Store struct {
+	embedder embeddings.Embedder
+
+	indexName   string
+	projectName string
+	environment string
+	apiKey      string
+	textKey     string
+	nameSpace   string
+	useGRPC     bool
+}
+
+var _ vectorstores.VectorStore = Store{}
+
+// New crates a new Store with options.
+func New(opts ...Option) (Store, error) {
+	return applyClientOptions(opts...)
+}
+
+// AddDocuments creates vector embeddings from the documents using the embedder
+// and upsert the vectors to the pinecone index.
+func (s Store) AddDocuments(ctx context.Context, docs []schema.Document) error {
+	texts := make([]string, 0, len(docs))
+	for _, doc := range docs {
+		texts = append(texts, doc.PageContent)
+	}
+
+	vectors, err := s.embedder.EmbedDocuments(ctx, texts)
+	if err != nil {
+		return err
+	}
+
+	if len(vectors) != len(docs) {
+		return ErrEmbedderWrongNumberVectors
+	}
+
+	metadatas := make([]map[string]any, 0, len(docs))
+	for i := 0; i < len(docs); i++ {
+		metadata := make(map[string]any, len(docs[i].Metadata))
+		for key, value := range docs[i].Metadata {
+			metadata[key] = value
+		}
+		metadata[s.textKey] = texts[i]
+
+		metadatas = append(metadatas, metadata)
+	}
+
+	if s.useGRPC {
+		conn, err := grpcapi.GetGRPCConn(ctx, s.indexName, s.projectName, s.environment, s.apiKey)
+		if err != nil {
+			return err
+		}
+
+		return grpcapi.Upsert(ctx, conn, vectors, metadatas, s.nameSpace)
+	}
+
+	return restapi.Upsert(
+		ctx,
+		vectors,
+		metadatas,
+		s.apiKey,
+		s.nameSpace,
+		s.indexName,
+		s.projectName,
+		s.environment,
+	)
+}
+
+// SimilaritySearch creates a vector embedding from the query using the embedder
+// and queries to find the most similar documents.
+func (s Store) SimilaritySearch(ctx context.Context, query string, numDocuments int) ([]schema.Document, error) {
+	vector, err := s.embedder.EmbedQuery(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+
+	if s.useGRPC {
+		conn, err := grpcapi.GetGRPCConn(ctx, s.indexName, s.projectName, s.environment, s.apiKey)
```

Comment:
it's atypical in grpc to create a new connection for each rpc -> given that you said this isn't working yet can we pull this out and we can consider how to re-add?